### PR TITLE
Core application

### PR DESCRIPTION
**Pull request by jodosha**
## Features

**TL;DR** This is a rewriting of the core of Hanami

  - [x] New application
  - [x] New configuration syntax with defaults
  - [x] Internal container
  - [x] Unified routes
  - [x] Delegate code reloading to `hanami-reloader`

### DRY-up settings

The big change from Hanami 1 is the removal of all the applications (`web`, `admin`) as a way to micro-configure areas of your project.

In the past we had:

  * `config/application.rb`
  * `apps/{web,admin,...}/application.rb`

To configure where to load files from, which was the default response format for each single app, the sessions, cookies, base URL, just to name a few.

The concept is to **unify** all these settings at the level of `config/environment.rb`, so they are **propagated** to the apps. DRY at its finest.

### No more applications to hold settings

In order to not overwhelm newcomers, `apps/{web,admin,...}/application.rb` are gone.

Single apps can override the inherited settings from the main app, by tweaking settings in new concrete superclasses like `Web::Action` (`apps/web/action.rb`).

```ruby
# frozen_string_literal: true

require_relative "./actions/authentication"

module Web
  class Action < Hanami::Action
    include Actions::Authentication
  end
end
```

### Example

To make easier the transition from Rails, application term now refers to the main project. Here's an example of  `config/environment.rb`:

```ruby
# frozen_string_literal: true

module Soundeck
  class Application < Hanami::Application
    config.cookies  = { max_age: 600 }
    config.sessions = :cookie, { secret: ENV["SOUNDECK_SESSIONS_SECRET"] }
    config.middleware.use Middleware::Elapsed, "1.0"

    config.environment(:production) do |c|
      c.base_url = ENV["SOUNDECK_BASE_URL"]
      c.logger = { level: :info, formatter: :json }
    end
  end
end
```

🔍  I'm considering to **rename** the file from `config/environment.rb` to `config/application.rb`. Please let me know what you think.

## Configuration

I introduced a new syntax for the application configuration based on **getters** and **setters**, instead of the DSL we had in Hanami 1.

#### Before

```ruby
cookies max_age: 600
```

#### After

```ruby
config.cookies = { max_age: 600 }
```

In some cases like this it's a little uglier (IMO), but it helps to simplify the internals, by explicitly separating the intent of reading a value from the intent of setting it.

`Hanami::Configuration` now performs **atomic/thread-safe operations** both at the read and write time.

### Defaults

Another difference from the past is that configuration now ships with defaults, instead of making everything explicit in the generated app.

Let me give you an example:

#### Before

This was an explicit setting of each application configuration.

```ruby
# apps/web/application.rb
security.x_frame_options = "DENY"
```

#### After

Now the configuration has that value as a default, so the generated code is empty:

```ruby
```

In case a developer needs to change the default, they will do it for the **whole app**:

```ruby
# config/environment.rb
config.security.x_frame_options = "sameorigin"
```

---

Again, the goal is to simplify and not overwhelm newcomers.

Implemented settings:

  * `environment`: to yield env based settings ( `config.environment(:production) { ... }`)
  * `base_url`: to set the base URL of the app (`config.base_url = "https://example.com"`)
  * `logger`: to set the logger options (`config.logger = { level: :info, format: :json }`)
  * `routes`: to set the path to routes file (`config.routes = "path/to/routes"`)
  * `cookies`: to set cookies options (`config.cookies = { max_age: 300 }`)
  * `sessions`: to set session options (`config.sessions = :cookie, { secret: "abc" }`)
  * `default_request_format`: to set the fallback for request format (aka MIME Type) (`config.default_request_format = :json`)
  * `default_response_format`: to set the default response format (aka MIME Type) (`config.default_response_format = :json`)
  * `middleware` to mount Rack middleware
  * `security` to set security settings (see below)
  * `inflections` to configure inflections (see below)

Security settings:

  * `x_frame_options`: defaults to `"deny"` (`config.security.x_frame_options = "sameorigin"`)
  * `x_content_type_options`: defaults to `"nosniff"` (`config.security.x_content_type_options = nil`)
  * `x_xss_protection`: defaults to `"1; mode=block"` (`config.security.x_xss_protection = "1"`)
  * `content_security_policy`: defaults to `"form-action 'self'; frame-ancestors 'self'; base-uri 'self'; default-src 'none'; script-src 'self'; connect-src 'self'; img-src 'self' https: data:; style-src 'self' 'unsafe-inline' https:; font-src 'self'; object-src 'none'; plugin-types application/pdf; child-src 'self'; frame-src 'self'; media-src 'self'"` (`config.security.content_security_policy[:style_src] += " https://my.cdn.example"` for adding another source) (`config.security.content_security_policy[:plugin_types] = nil` to override the settings)

Please note that `nil` values will have the effect of **NOT** sending the corresponding HTTP response header. From the examples above `X-Content-Type-Options` won't be sent.

Setting a `nil` value to `config.security.content_security_policy` element, it will remove it from the header. From the example above `plugin-types` **WON'T** be part of `Content-Security-Policy` value.

## Internal Container

I built a new **internal** container based on `dry-system`.

The potential of `dry-system` in Hanami context is to make possible for third-party plugins to register themselves and to configure the settings of an application.

It's of course thread-safe, for each component we can define dependencies. It's lazy: we can boot only a portion of the needed components, or we can tell it to boot up everything, which is useful when the app starts with the `hanami server` command.

It has for now the following components:

  * `root`: holds the absolute path to the root of the app
  * `env`: loads ENV vars from a `.env`
  * `lib`: recursively requires all the files under `lib/`
  * `environment`: requires `config/environment.rb`
  * `logger`: setup logger, according to the user settings
  * `routes`: requires `config/routes.rb` (new addition, see below)
  * `apps`: register applications defined in `config/routes.rb`
  * `actions`: requires `apps/{web,admin,...}/action.rb`, register action settings for each app
  * `code`: requires all the code from `apps`

## Removed `shotgun`

Removed `shotgun` as way to reload the code.

The `hanami` gem won't even know about code reloading anymore. `hanami new` will generate a `Gemfile` with `hanami-reloader` gem, which is able to handle code reloading from the **outside**.

This decision is made because of historical issues regarding `shotgun`:

  * It took years to find a fragile solution to make it to work in the Hanami context
  * It doesn't work with JRuby and `webconsole` gem
  * It was cluttering the internals of the framework

## Removed `.hanamirc`

Another step towards the simplification of the settings is the removal of `.hanamirc`. We used it to hold the following settings:

  * `project`: the project name, which can now be inferred from the `lib/` namespace
  * `template`: the template engine (e.g. `erb`), which we want delegate for the future to external gems
  * `testing`: the testing framework (e.g. `rspec`), same as `template`.

## Routes

Simplification again. In the past we had several places to configure the whole routes of an app: `apps/{web,admin,...}/config/routes.rb`. Now these files are gone in favor of `config/routes.rb` at the root of the project.

Thanks to `hanami-router` 2, we can define all the routes for all the apps in **a single shot**:

```ruby
# frozen_string_literal: true

Hanami.application.routes do
  mount :web, at: "/" do
    root to: "home#index"
  end

  mount :admin, at: "/admin" do
    root to: "home#index"
  end
end
```

Please note the differences from the past:

  * No more multiple files and multiple routers instances for multiple apps
  * Single router for the whole app 🙌 
  * Mounting apps is no longer delegated to the gone `Hanami.configure`, which was cluttering `config/environment.rb`
  * Apps as concrete classes are gone (e.g. `Web::Application`), in favor of symbols
  * The apps registered with `mount` are the ones that are activated in the internal container

## Inflections

Inflections are configured at the app level and they are propagated to all the components (mainly `router` and `model` in the future). We'll use `dry-inflector` for the purpose.

Here's an example:

```ruby
module Soundeck
  class Application < Hanami::Application
    config.inflections do |i|
      i.plural "virus", "viruses"
    end
  end
end
```

## Available commands

Because I'm rebuilding the core from the scratch, for this PR, only the following commands are available:

  * `hanami version`
  * `hanami server`

## How to test it

I created a testing app so you can see how Hanami 2 is taking shape: https://github.com/jodosha/soundeck

### CODE REVIEW

**Comment by solnic** [comment]
> 🔍 I'm considering to rename the file from config/environment.rb to config/application.rb. Please let me know what you think.

+💯 environment is too ambiguous IMO, whereas application says what it is

**Comment by sandelius** [comment]
Wow, this a huge change. I kinda like the whole container architecture we have now tho. But I can see how that can be a bit overwhelming for newcomers.

**Comment by jdickey** [comment]
One of the things that first drew me to Hanami was its approach to "monolith first". Yes, start your app as a monolith, but have the framework make it trivial (in the 1.x versions) to extract logic first to a separate app under `apps/whatever` and then, if desired, to an entirely separate codebase/Gem that the "original" app used.

By simplifying things for the Rails refugees (and I understand them; I was one recently), how is that ability going to be affected? Of course it's not used heavily _yet_; Hanami is still a new framework seen as "risky" and "experimental" in a lot of Rails shops. But by encouraging that aspect of Clean Architecture, Hanami was making promises about a cleaner, more maintainable future once large apps started to switch over to it. Whither Clean Architecture now? Or am I fundamentally misunderstanding things again?

**Comment by cllns** [comment]
> One of the things that first drew me to Hanami was its approach to "monolith first". Yes, start your app as a monolith, but have the framework make it trivial (in the 1.x versions) to extract logic first to a separate app under `apps/whatever` and then, if desired, to an entirely separate codebase/Gem that the "original" app used.
> 
> By simplifying things for the Rails refugees (and I understand them; I was one recently), how is that ability going to be affected? Of course it's not used heavily _yet_; Hanami is still a new framework seen as "risky" and "experimental" in a lot of Rails shops. But by encouraging that aspect of Clean Architecture, Hanami was making promises about a cleaner, more maintainable future once large apps started to switch over to it. Whither Clean Architecture now? Or am I fundamentally misunderstanding things again?

The multiple apps still exist, but the _configuration_ for the apps is unified into a single-file. It streamlines things.

**Comment by jdickey** [comment]
@cllns So, say, extracting one app into a separate Gem (for reuse with other related apps) would be pretty much as with 1.3, then? The only thing touched in the pre-existing app would be the appropriate bits of config ripped out and copied over, and the Gem hooked in to routes and such? That's certainly still cleaner than it could have been, or that I misunderstood it as being. Thanks for the clarification. I'll have to watch Luca's video again.

File: lib/hanami/routes.rb [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+      @apps = []
+      instance_eval(&blk)
+    end
+
+    def to_proc
+      @blk
+    end
+
+    def mount(app, **)
+      @apps << app if app.is_a?(Symbol)
```
Comment (timriley) [inline lib/hanami/routes.rb]: Feels like this should fail more noisily if it's ever passed something besides a symbol?

File: lib/hanami.rb [hunk 2]
```diff
@@ -8,4 +10,64 @@
...
+
+  def self.application=(app)
+    @_mutex.synchronize do
+      raise "Hanami application already configured" if defined?(@_application)
+
+      @_application = app unless app.name.nil?
+    end
+  end
+
+  def self.app
```
Comment (timriley) [inline lib/hanami.rb]: I wonder if we want a different name for this, since the pairing of `application`/`app` may be a little confusing?

File: lib/hanami/container.rb [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+      init do
+        use :root
+        use :configuration
+        use :apps
+        use :actions
+
+        apps = c[:apps].join(",")
+        Hanami::Utils.require!(c[:root].join("apps", "{#{apps}}", "**", "*.rb"))
+      end
+    end
```
[the remark this answers is not in the document]
Comment (jodosha) [reply lib/hanami/container.rb]: @timriley This sounds like a nice to have feature, but without a real, immediate demand for it. I'd pass it for now.

File: lib/hanami/routes.rb (hunk 1 above)
Comment (jodosha) [reply lib/hanami/routes.rb]: @timriley Please consider that this is **NOT** exposed directly to users. This is a trick to extract the Hanami _slices_ mounted inside `config/router.rb`.

However:

  1. The object tha powers directly `config/routes.rb` is `Hanami::Router#mount` redefined by `lib/hanami/frameworks.rb` in this PR.
  2. That `#mount` API must allow non-symbol arguments to mount Rack apps. (e.g. `mount MySinatraApp.new`)

File: lib/hanami.rb (hunk 2 above)
Comment (jodosha) [reply lib/hanami.rb]: @timriley This **is** the only instance of `Hanami::Application` subclass in the process. So it's **the** app.

```ruby
Hanami.app 
# => #<Soundeck::Application:0x00007ffee940eee0 ...>
```

**Comment by jodosha** [comment]
@timriley Regarding `.hanamirc`, won't be implemented anymore. It isn't part of this PR, but the information can be inferred like this:

  1. Project/App name: from the `Hanami::Application` subclass. Eg. `Soundeck::Application`, per convention, `"soundeck"` will be inferred from there.
  1. Template framework: from a setting that can be either in the view framework or in the project configuration
  1. Testing framework: I want to extract `hanami-rspec` and `hanami-minitest`, which will register themselves as testing framework and will be able to generate the proper testing files.

File: lib/hanami/container.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (timriley) [reply lib/hanami/container.rb]: @jodosha I'm fine if you want to pass it for now (i.e. for this PR), but I think we should have a discussion about this before 2.0 settles in full. From actual practice (i.e. several quite large apps here at Icelab), having separate containers turns out to be a really helpful way to arrange/separate the components for distinct sections (i.e. slices) within a larger application and you're probably going to have to reimplement some of the resulting behaviour yourself when you work out selective booting of the slices.

Perhaps we can discuss sometime (along with @solnic or @flash-gordon) after this PR is merged?

File: lib/hanami/routes.rb (hunk 1 above)
Comment (timriley) [reply lib/hanami/routes.rb]: @jodosha OK, with this extra info it makes more sense to me how this particular method works, thanks.

So this particular Hanami::Routes class has 2 key jobs: (a) holds the block that is passed when it is initialized (so it can be passed to the real router later on?) and (b) holds a set of apps/slices, so the app can know what has been mounted? Is this right?

Comment (timriley) [reply lib/hanami/routes.rb]: What's still not clear is why we need this `method_missing(*)` implementation?

File: lib/hanami.rb (hunk 2 above)
Comment (timriley) [reply lib/hanami.rb]: Ah, so for a booted application, `Hanami.application` is the application's _class_ (via the `Hanami::Application.inherited` hook) and `Hanami.app` is the one _instance_ of that application class? Is that right? i.e.

```ruby
Hanami.application
# => Soundeck::Application

Hanami.app
# => #<Soundeck::Application:0x00007ffee940eee0 ...>
```

I still feel like that's a pretty confusing pairing of names. "app" is just shorthand for "application", they're synonyms. It feels a little odd that we use them both here, holding 2 different (though obviously related) things.

File: lib/hanami/container.rb [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+    boot(:env) do |c|
+      init do
+        use :root
+
+        begin
+          require "dotenv"
+        rescue LoadError # rubocop:disable Lint/HandleExceptions
+        end
+
+        Dotenv.load(c[:root].join(".env")) if defined?(Dotenv)
```
[the remark this answers is not in the document]
Comment (timriley) [reply lib/hanami/container.rb]: Having pluggable settings providers certainly _sounds_ like a nice thing to me. I guess it'll come down to whether we can maintain a consistent settings interface/experience while still allowing each settings provider's unique features to be leveraged.

What I'm definitely in favour of is having the settings explicitly listed (ideally along with their types), and then vended from a central object, which is what that dry-system settings component does. It's much cleaner than ENV and opens up possibilities for doing smarter things with settings, since we have a central place to manage them.

Happy for this to be a follow-up issue.

File: lib/hanami/container.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (solnic) [reply lib/hanami/container.rb]: Yes, having multiple containers per slice and giving them access to the "main" container so that they can import core components (if needed) is a great architecture that was already proven to work well in practice. Making it work like that from the beginning will certainly reduce the risk of having to do bigger'ish refactorings later on. It doesn't have to be a part of this PR but it'd be great to address this prior 2.0 final. I'm happy to help with this.

[the remark this answers is not in the document]
Comment (solnic) [reply lib/hanami/container.rb]: Blah sorry, I meant *having multiple containers per app where each slice has its own container*, of course 😄

File: lib/hanami.rb (hunk 2 above)
Comment (jodosha) [reply lib/hanami.rb]: @timriley Which name do you suggest? 😃 Please remember that `Hanami.app` is short and memorable, and it's used by public facing integration specs.

Eg.

```ruby
require "features_helper"

RSpec.feature "Signup" do
  let(:app) { Hanami.app } # for capybara and rack-test
end
```

File: lib/hanami/container.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (jodosha) [reply lib/hanami/container.rb]: Oh, that's interesting. This doesn't belong really to this PR. We'll need to discuss about this.

File: lib/hanami/container.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (jodosha) [reply lib/hanami/container.rb]: @timriley Is that DRY mechanism that you linked a kind of replacement for `dotenv`? If so, does it support multiple envs and/or overrides via local files?

File: lib/hanami.rb (hunk 2 above)
Comment (cllns) [reply lib/hanami.rb]: @jodosha What about `project` for the container application? You used "project" four times in the main message to refer to this :) It's not quite as memorible or short as `app`, but it's not hard to remember either. I agree with @timriley that using `app` and `application` as different things would be confusing.

File: lib/hanami/cli/commands.rb [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+    #
+    # @example Third party gem
+    #   require "hanami/cli/commands"
+    #
+    #   module Hanami
+    #     module Webpack
+    #       module CLI
+    #         module Commands
+    #           class Generate < Hanami::CLI::Command
+    #             desc "Generate Webpack config"
```
Comment (kaikuchn) [inline lib/hanami/cli/commands.rb]:
```suggestion
    #             desc "Generate Webpack configuration"
```

I assume this is the string that should match the one in line 48, right?

File: lib/hanami.rb (hunk 2 above)
Comment (jodosha) [reply lib/hanami.rb]: I'm not sure to if keeping project is a good idea. The intention is to remove the word "project" from the terminology. See my YT video.

File: lib/hanami/cli/commands.rb (hunk 5 above)
Comment (jodosha) [reply lib/hanami/cli/commands.rb]: @kaikuchn That's correct, good catch 👍

File: lib/hanami/configuration.rb [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+      self.default_response_format = DEFAULT_RESPONSE_FORMAT
+
+      self.middleware = Middleware.new
+      self.security   = Security.new
+
+      self.inflections = Dry::Inflector.new
+    end
+    # rubocop:enable Metrics/MethodLength
+
+    def finalize
```
[the remark this answers is not in the document]
Comment (jodosha) [reply lib/hanami/configuration.rb]: Because `Hanami::Container` is subclassing `Dry::System::Container`, which exposes `.finalize` as a booting trigger, for consistency sake, I named this method the same.

Given this is a private method, I don't think people will look at it, nor confuse with Ruby's object finalization.

File: lib/hanami/cli/commands/server.rb [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+            Host: host(args),
+            Port: port(args),
+            AccessLog: []
+          }
+
+          OPTIONAL_SETTINGS.each do |setting|
+            next unless args.key?(setting)
+
+            result[setting] = args.fetch(setting)
+          end
```
[the remark this answers is not in the document]
Comment (jodosha) [reply lib/hanami/cli/commands/server.rb]: Great suggestion. Changed by 684d9f6

File: lib/hanami/configuration.rb [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+    # rubocop:disable Metrics/MethodLength
+    def initialize(env:)
+      @settings = Concurrent::Hash.new
+
+      self.env = env
+      self.environments = DEFAULT_ENVIRONMENTS.dup
+
+      self.base_url = DEFAULT_BASE_URL
+
+      self.logger   = DEFAULT_LOGGER.dup
```
[the remark this answers is not in the document]
Comment (jodosha) [reply lib/hanami/configuration.rb]: Really good catch. 👍 Fixed by 3ab512c

File: lib/hanami/configuration.rb [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+
+    def inflections(&blk)
+      if blk.nil?
+        settings.fetch(:inflections)
+      else
+        settings[:inflections] = Dry::Inflector.new(&blk)
+      end
+    end
+
+    def to_router
```
[the remark this answers is not in the document]
Comment (jodosha) [reply lib/hanami/configuration.rb]: Changed by ed2f920

File: lib/hanami/configuration/security.rb [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+        settings.fetch(:content_security_policy)
+      end
+
+      def to_hash
+        settings.each_with_object({}) do |(_, v), result|
+          next if v.value.nil?
+
+          result[v.header] = v.value
+        end
+      end
```
[the remark this answers is not in the document]
Comment (jodosha) [reply lib/hanami/configuration/security.rb]: In this way we skip the `O(n)` of `Hash#compact` and avoid the creation of an intermediate Hash.

File: lib/hanami/configuration/security.rb [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+
+      # Content security policy settings
+      #
+      # @api private
+      # @since 2.0.0
+      class ContentSecurityPolicy < Setting
+        def value
+          @value.map do |k, v|
+            "#{Utils::String.dasherize(k)} #{v}" unless v.nil?
+          end.compact.join("; ")
```
[the remark this answers is not in the document]
Comment (jodosha) [reply lib/hanami/configuration/security.rb]: Same as https://github.com/hanami/hanami/pull/975#discussion_r246375224

File: lib/hanami/configuration/sessions.rb [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+      end
+
+      attr_reader :storage, :options
+
+      def initialize(*args)
+        args = Array(args.dup).flatten
+
+        @storage = args.shift
+        @options = args.shift || {}
+      end
```
[the remark this answers is not in the document]
Comment (jodosha) [reply lib/hanami/configuration/sessions.rb]: @kaikuchn That permissive signature is used to make possible the syntax `config.sessions = :cookie, { secret: "..." }`.

In 257e988 I changed the assigment as suggested by you, so it makes clear what we expect to receive in `*args`. 👍

File: lib/hanami/container.rb [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+      init do
+        use :configuration
+        use :apps
+
+        c[:apps].each do |app|
+          require c[:root].join("apps", app.to_s, "action")
+
+          namespace = Utils::String.classify("#{app}::Actions")
+          namespace = Utils::Class.load!(namespace)
+          # action    = Utils::Class.load!("#{namespace}::Action")
```
[the remark this answers is not in the document]
Comment (jodosha) [reply lib/hanami/container.rb]: Removed by d9566d3

**Comment by jodosha** [comment]
@kaikuchn Thanks for your feedback. I addressed all the feedback points. 👍

File: lib/hanami/container.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (timriley) [reply lib/hanami/container.rb]: @jodosha it supports environment-named files, but not the `.local` versions. We could extend it to include those, but perhaps a better option to have this dry-system plugin somehow abstract away the settings provider (like we discussed just above, here)

File: lib/hanami.rb (hunk 2 above)
Comment (timriley) [reply lib/hanami.rb]: Maybe we could stick with app in both methods but disambiguate between the two with a suffix, e.g. `#app` and `#app_class`?

File: lib/hanami/container.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (solnic) [reply lib/hanami/container.rb]: @jodosha @timriley loading of env files is currently hard-coded [here](https://github.com/dry-rb/dry-system/blob/master/lib/dry/system/settings/file_loader.rb#L19-L24) but it would be very, very simple to add a configuration to `:settings` component that would specify file name hierarchy.

File: lib/hanami.rb (hunk 2 above)
Comment (solnic) [reply lib/hanami.rb]: Given that `app` returns an instance of the application, I would say a natural name for the other method should be `app_class`. I would also say this should be a private API anyway.

File: lib/hanami/container.rb [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+module Hanami
+  # Hanami private IoC
+  #
+  # @since 2.0.0
+  class Container < Dry::System::Container
+    boot(:root) do
+      start do
+        register(:root, Hanami.root)
+      end
+    end
```
Comment (solnic) [inline lib/hanami/container.rb]: This is just a configuration, `root` is not a component. What's the reason for treating it as a component? Furthermore, dry-system's containers have `config.root` which you want to set correctly.

File: lib/hanami/container.rb [hunk 15]
```diff
@@ -0,0 +1,115 @@
...
+
+    boot(:lib) do |c|
+      init do
+        use :root
+
+        $LOAD_PATH.unshift root.join("lib")
+        Hanami::Utils.require!(c[:root].join("lib", "**", "*.rb"))
+      end
+    end
```
Comment (solnic) [inline lib/hanami/container.rb]: This shouldn't be needed. Why blindly requiring all the lib files?

File: lib/hanami/container.rb [hunk 16]
```diff
@@ -0,0 +1,115 @@
...
+      init do
+        use :root
+
+        require c[:root].join("config", "application").to_s
+      end
+
+      start do
+        register(:configuration, Hanami.application.configuration.finalize)
+      end
+    end
```
Comment (solnic) [inline lib/hanami/container.rb]: Why `configuration` is stored in the class instead of the instance?

File: lib/hanami/container.rb [hunk 17]
```diff
@@ -0,0 +1,115 @@
...
+
+    boot(:code) do |c|
+      init do
+        use :root
+        use :configuration
+        use :apps
+        use :actions
+
+        apps = c[:apps].join(",")
+        Hanami::Utils.require!(c[:root].join("apps", "{#{apps}}", "**", "*.rb"))
```
[the remark this answers is not in the document]
Comment (solnic) [reply lib/hanami/container.rb]: @davydovanton it is done in the `lib` component but I pointed out it doesn't sound like a good idea to me

[the remark this answers is not in the document]
Comment (timriley) [reply lib/hanami/container.rb]: (@solnic must have typed his comments about this while I was writing my one out!)

File: lib/hanami/container.rb [hunk 18]
```diff
@@ -0,0 +1,115 @@
+# frozen_string_literal: true
+
+require "dry/system/container"
+
+module Hanami
+  # Hanami private IoC
+  #
+  # @since 2.0.0
+  class Container < Dry::System::Container
+    boot(:root) do
```
[the remark this answers is not in the document]
Comment (timriley) [reply lib/hanami/container.rb]: Another idea would be to have a whole framework-specific _container_, which then gets imported from application and slice containers, this would take care of the namespacing and also help encourage better separation of the code within the various broad sections of the app

[the remark this answers is not in the document]
Comment (davydovanton) [reply lib/hanami/container.rb]: Hmm. @timriley, you give me an idea. I think, it will be awesome to put `Hanami.boot` to general container and generate `./system/` folder as a part of framework. And in this case instead this code:
```
# frozen_string_literal: true

require_relative './environment'
Hanami.boot
```

use something like this:

```
Hanami::Container.boot!
```

And also, anyone will can use dry-system booting process instead `./config/initializers/*.rb` magic 🤔

[the remark this answers is not in the document]
Comment (timriley) [reply lib/hanami/container.rb]: oh yes, we should be 100% offering dry-system bootable components instead of random initialisers!

[the remark this answers is not in the document]
Comment (timriley) [reply lib/hanami/container.rb]: @davydovanton perhaps you could demonstrate what this change would look like, perhaps a diff based on this PR?

File: lib/hanami/container.rb (hunk 17 above)
[the remark this answers is not in the document]
Comment (davydovanton) [reply lib/hanami/container.rb]: @solnic got it, thanks! agree that `lib` looks strange here but as a starting point  it's good for me 👍 

@timriley 💯 for your idea. In this case, we can drop a lot of code for loading and drop unnecessary and strange part of "ruby" frameworks (initializers for example). Also, not sure but I think using one container for booting all dependencies from project allow us to use better reload and in future we can add "partial rebooting" to dry-system 🤔 

Something like: if we update `users` repository in dev mode, we can update only `Container['core.repositories.users']` key without any unnecessary reloading other parts of the system

File: lib/hanami/container.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (jodosha) [reply lib/hanami/container.rb]: @timriley & I had a long discussion regarding settings and env vars. We decided to leave the PR as it is, and to structure this feature in a future PR.

File: lib/hanami/container.rb (hunk 18 above)
[the remark this answers is not in the document]
Comment (jodosha) [reply lib/hanami/container.rb]: @davydovanton Given we're not sure if we're gonna have a separate place for settings, or multiple containers, I'd leave this as it is.

We need to address this aspect soon.

File: lib/hanami/container.rb (hunk 17 above)
[the remark this answers is not in the document]
Comment (jodosha) [reply lib/hanami/container.rb]: @davydovanton We require it here: https://github.com/hanami/hanami/pull/975/files#diff-1ed1deb125747568d4b8c969549060e5R29

File: lib/hanami/container.rb (hunk 18 above)
[the remark this answers is not in the document]
Comment (davydovanton) [reply lib/hanami/container.rb]: @jodosha agree, I can create proposal and we can start discussion. WDYT?

File: lib/hanami.rb (hunk 2 above)
Comment (cllns) [reply lib/hanami.rb]: I can understand some of the confusion between "Hanami project" referring to this whole effort, but I still think "project" is a natural and obvious way to talk about what the "umbrella" app is.

Will `apps/` be renamed to `slices/`? They're still Rack-compatible apps. While I like the naming of "slices" and it seems appropriate, I feel like this is re-inventing the wheel and causes more confusion than it's worth. `Hanami.app` (or `Hanami.application`) being a different thing from `Hanami::App` is confusing. If we want to use "slice", we should go all the way and rename `Hanami::App` to `Hanami::Slice`. I don't think we should do that, but we should be consistent if we're changing the terminology.

Django uses the `project`/ `application` distinction in the same way we have been: https://medium.com/@timmykko/a-quick-glance-of-django-for-beginners-688bc6630fab
> Any Django _project_ will have at least one Django _app_. Yes, I know, it’s a little confusing. A Django _project_ encompasses the application and all its components, while a Django _app_ is a sub-container of the application with its own functionality that, in theory, may be reusable in another application without much modification.


(Also, for others, [here's the YouTube video](https://www.youtube.com/watch?v=mQCTNswIhl8), since it hasn't been linked in this PR yet)

File: lib/hanami/container.rb (hunk 16 above)
Comment (jodosha) [reply lib/hanami/container.rb]: It's stored at the class level to allow the _config DSL_. Example:

```ruby
module Soundeck
  class Application < Hanami::Application
    # This corresponds to Hanami.application.configuration
    config.whatever = :value
  end
end
```

As you know, I'm trying my best to remove class configurations and global state, but some glue code has still exist in a few spots to make it work.

Comment (solnic) [reply lib/hanami/container.rb]: > It's stored at the class level to allow the config DSL

Ah ok, so is it `config` vs `configuration`? Maybe this should be always `config`.

> As you know, I'm trying my best to remove class configurations and global state, but some glue code has still exist in a few spots to make it work.

That's understandable!

Comment (jodosha) [reply lib/hanami/container.rb]: It's mainly `configuration`, but it's aliased to save some keystrokes to devs. https://github.com/hanami/hanami/pull/975/files#diff-00f27089d3c7dbf5783b4a35405b603aR36

File: lib/hanami/container.rb (hunk 14 above)
Comment (jodosha) [reply lib/hanami/container.rb]: I changed it to use the appropriate container setting: https://github.com/hanami/hanami/pull/975/commits/cfd7adbc2d996bc076dcec9720fdf50912c78929

File: lib/hanami/container.rb (hunk 15 above)
Comment (jodosha) [reply lib/hanami/container.rb]: This is for convenience sake. Given under `lib/` there are repositories and entities, that are necessary for the app.

Do you think [_Auto-Registration_](https://dry-rb.org/gems/dry-system/container/) should be used here instead? Or do you suggest to let devs to manually require code?

File: lib/hanami.rb (hunk 2 above)
Comment (cllns) [reply lib/hanami.rb]: @jodosha Thoughts on the above? ☝️  (Don't mean to bother you, I just want to make sure you saw it since I didn't mention you in the comment.) 🌸

File: lib/hanami/container.rb (hunk 15 above)
Comment (solnic) [reply lib/hanami/container.rb]: @jodosha sorry for replying late - *my* preference is to "require what you require", with no exceptions. Probably a good solution in Hanami would be to offer "load everything" as an option that users can turn on/off. In such case folks coming from Rails would feel more comfortable when it's turned on.

> Do you think Auto-Registration should be used here instead?

Partly yes. Auto-registration works great for all objects that are used as injectible dependencies. There are exceptions though - ie entities are not used in DI, so they shouldn't be auto-registered.
